src/bson.rs

```diff
@@ -1054,6 +1057,118 @@ impl Binary {
+/// UUID, with the BSON subtype: 0x03 UUID old.
+/// If a UUID has been serialized with a particular representation, it MUST
+/// be deserialized with the same representation.
+///
+/// Example:
+/// ```
+/// use crate::{bson::UuidRepresentation, bson::Binary};
+/// use uuid::Uuid;
+/// let uuid = Uuid::parse_str("00112233445566778899AABBCCDDEEFF").unwrap();
+/// let bin = Binary::from_uuid_with_representation(uuid, UuidRepresentation::PythonLegacy);
+/// let new_uuid = bin.to_uuid();
+/// assert!(new_uuid.is_err());
+/// let new_uuid = bin.to_uuid_with_representation(UuidRepresentation::PythonLegacy);
+/// assert!(new_uuid.is_ok());
+/// assert_eq!(new_uuid.unwrap(), uuid);
+/// ```
+#[non_exhaustive]
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum UuidRepresentation {
+    /// The canonical representation of UUID BSON binary subtype 4
+    Standard,
+    /// The legacy representation of UUID used by the C# driver BSON binary subtype 3
+    CSharpLegacy,
+    /// The legacy representation of UUID used by the Java driver BSON binary subtype 3
+    JavaLegacy,
+    /// The legacy representation of UUID used by the Python driver, which is the same format as STANDARD, but has the UUID old BSON subtype (\x03) BSON binary subtype 3
+    PythonLegacy,
+}
+
+#[cfg(feature = "uuid-0_8")]
+#[cfg_attr(docsrs, doc(cfg(feature = "uuid-0_8")))]
+impl Binary {
+    pub fn from_uuid(uuid: uuid::Uuid) -> Self {
+        Binary::from(uuid)
+    }
+
+    pub fn from_uuid_with_representation(uuid: uuid::Uuid, rep: UuidRepresentation) -> Self {
+        match rep {
+            UuidRepresentation::Standard => Binary::from_uuid(uuid),
+            UuidRepresentation::CSharpLegacy => {
+                let mut bytes = uuid.as_bytes().to_vec();
+                bytes[0..4].reverse();
+                bytes[4..6].reverse();
+                bytes[6..8].reverse();
+                Binary {
+                    subtype: BinarySubtype::UuidOld,
+                    bytes,
+                }
+            }
+            UuidRepresentation::PythonLegacy => Binary {
+                subtype: BinarySubtype::UuidOld,
+                bytes: uuid.as_bytes().to_vec(),
+            },
+            UuidRepresentation::JavaLegacy => {
+                let mut bytes = uuid.as_bytes().to_vec();
+                bytes[0..8].reverse();
+                bytes[8..16].reverse();
+                Binary {
+                    subtype: BinarySubtype::UuidOld,
+                    bytes,
+                }
+            }
+        }
+    }
+
+    pub fn to_uuid_with_representation(
+        &self,
+        rep: UuidRepresentation,
+    ) -> Result<uuid::Uuid, crate::de::Error> {
+        // If representation is non-standard, then its subtype must be UuidOld
+        if rep != UuidRepresentation::Standard && self.subtype != BinarySubtype::UuidOld {
+            return Err(Error::custom(
+                "expecting binary subtype 3 for non-standard representations",
```

Comment:
can we add some more context to this error? It'll make it easier for users to debug.

e.g.
```rust
format!("expected binary subtype 3 when converting to UUID, instead got {:#04x}", u8::from(self.subtype))
```